File: python/2015/day04a/day04a.py
#!/usr/bin/env python3

import hashlib
import sys

def usage():
    progname = sys.argv[0]
    print(f'usage: {progname} <key>')
    sys.exit(1)

def process(key):
    n = 1
    while True:
        tryKey = key + str(n)
        md5 = hashlib.md5()
        md5.update(tryKey.encode('utf-8'))
        digest = md5.hexdigest()
        if digest.startswith('00000'):
            return n
        n += 1

def main():
    if len(sys.argv) < 2:
        usage()
    key = sys.argv[1]
    result = process(key)
    print(f'result = {result}')

if __name__ == '__main__':
    main()

File: dart/2015/day04a/day04a.dart
#!/usr/bin/env dart

import 'dart:convert';
import 'dart:io';
import 'package:crypto/crypto.dart';

void usage(String progName) {
    print('usage: $progName <key>');
    exit(1);
}

int process(String key) {
    int n = 1;
    while ( true ) {
        String tryKey = key + n.toString();
        final bytes = utf8.encode(tryKey);
        final digest = md5.convert(bytes);
        String hexDigest = digest.toString();
        if ( hexDigest.startsWith("00000") ) {
            return n;
        }
        n++;
    }
}

void main(List<String> args) {
    if ( args.isEmpty ) {
        usage(Platform.script.pathSegments.last);
    }
    final key = args[0];
    final result = process(key);
    print('result = $result');
}
